### PR TITLE
feat: Add TreatmentEffect and TreatmentEffectBuilder 

### PR DESCRIPTION
**Pull request by LauraBoenchenLB**

### CODE REVIEW

File: medmodels/treatment_effect_estimation/analysis_modules/estimate.py [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+        """Sorts patients into the contingency table of treatment-outcome, treatment-
+            no outcome, control-outcome and control-no outcome. The treatment group and
+            control matching is determined based on the treatment effect configuration.
+
+        Args:
+            medrecord (MedRecord): The MedRecord object containing the data.
+
+        Returns:
+            Tuple[Set[NodeIndex], Set[NodeIndex], Set[NodeIndex], Set[NodeIndex]: The
+                patient ids of true and false subjects in the treatment and control groups, respectively.
```
Comment (JabobKrauskopf) [inline medmodels/treatment_effect_estimation/analysis_modules/estimate.py:61]: Line too long

File: medmodels/treatment_effect_estimation/analysis_modules/estimate.py [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+        )
+        assert self._treatment_effect._outcomes_group in medrecord.groups, (
+            "Outcome group not found in the data. "
+            f"Available groups: {medrecord.groups}"
+        )
+
+    def _sort_patients_in_contingency_table(
+        self, medrecord: MedRecord
+    ) -> Tuple[Set[NodeIndex], Set[NodeIndex], Set[NodeIndex], Set[NodeIndex]]:
+        """Sorts patients into the contingency table of treatment-outcome, treatment-
```
Comment (JabobKrauskopf) [inline medmodels/treatment_effect_estimation/analysis_modules/estimate.py:52]: The format for the docstrings is not consistent with the rest of the codebase. We usually start in the next line

File: medmodels/treatment_effect_estimation/analysis_modules/estimate.py [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+    ) -> Dict[str, Set[NodeIndex]]:
+        """
+        Overview of which patients were used for the treatment effect.
+
+        Args:
+            medrecord (MedRecord): The MedRecord object containing the data.
+
+        Returns:
+            Dict[str, Set[NodeIndex]]: Dictionary with description of the subject group
+            and Lists of patient ids belonging to each group.
```
Comment (JabobKrauskopf) [inline medmodels/treatment_effect_estimation/analysis_modules/estimate.py:147]: This docstring is missing the Raises: AssertionError

File: medmodels/treatment_effect_estimation/analysis_modules/estimate.py [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+
+    def subject_counts(self, medrecord: MedRecord) -> Dict[str, int]:
+        """
+        Returns the subject counts for the treatment and control groups in a
+        Dictionary.
+
+        Args:
+            medrecord (MedRecord): The MedRecord object containing the data.
+
+        Returns:
```
Comment (JabobKrauskopf) [inline medmodels/treatment_effect_estimation/analysis_modules/estimate.py:171]: This docstring is missing the Raises: AssertionError

File: medmodels/treatment_effect_estimation/analysis_modules/estimate.py [hunk 5]
```diff
@@ -0,0 +1,407 @@
...
+            float: The calculated relative risk between the treatment and control
+                groups.
+
+        Raises:
+            AssertionError: If the preconditions are not met, indicating a potential
```
Comment (JabobKrauskopf) [inline medmodels/treatment_effect_estimation/analysis_modules/estimate.py:214]: Should probably be the same Raises docstring as the other functions. Or at least not so ambiguous

File: medmodels/treatment_effect_estimation/analysis_modules/estimate.py [hunk 6]
```diff
@@ -0,0 +1,407 @@
...
+                calculations.
+
+        Args:
+            medrecord (MedRecord): The MedRecord object containing the data.
+
+        Returns:
+            float: The calculated odds ratio between the treatment and control groups.
+
+
+        Raises:
```
Comment (JabobKrauskopf) [inline medmodels/treatment_effect_estimation/analysis_modules/estimate.py:252]: Should probably be the same Raises docstring as the other functions. Or at least not so ambiguous

File: medmodels/treatment_effect_estimation/analysis_modules/estimate.py [hunk 7]
```diff
@@ -0,0 +1,407 @@
...
+        influenced by factors other than the treatment.
+
+        Precondition:
+            - Subject counts in each group must be non-zero to avoid division by zero
+                errors.
+
+        Args:
+            medrecord (MedRecord): The MedRecord object containing the data.
+
+        Returns:
```
Comment (JabobKrauskopf) [inline medmodels/treatment_effect_estimation/analysis_modules/estimate.py:291]: Missing raises

File: medmodels/treatment_effect_estimation/analysis_modules/estimate.py [hunk 8]
```diff
@@ -0,0 +1,407 @@
...
+        return numerator / denominator
+
+    def absolute_risk(self, medrecord: MedRecord) -> float:
+        """
+        Calculates the absolute risk (AR) of an event occurring in the treatment group
+        compared to the control group. AR is a measure of the incidence of an event in
+        each group.
+
+
+        Preconditions:
```
Comment (JabobKrauskopf) [inline medmodels/treatment_effect_estimation/analysis_modules/estimate.py:322]: I don't think preconditions are part of the google python docstring styleguide. This should probably just be handled in the Raises part

Comment (JabobKrauskopf) [reply medmodels/treatment_effect_estimation/analysis_modules/estimate.py:322]: also applies to some other functions in this file

File: medmodels/treatment_effect_estimation/analysis_modules/estimate.py [hunk 9]
```diff
@@ -0,0 +1,407 @@
...
+        """
+        (
+            num_treat_true,
+            num_treat_false,
+            num_control_true,
+            num_control_false,
+        ) = self._compute_subject_counts(medrecord=medrecord)
+
+        return (num_treat_true / (num_treat_true + num_treat_false)) / (
+            num_control_true / (num_control_true + num_control_false)
```
Comment (JabobKrauskopf) [inline medmodels/treatment_effect_estimation/analysis_modules/estimate.py:225]: In some functions in this file there are assertions before possibly dividing by zero, in some cases there aren't. This should be consistent

File: medmodels/treatment_effect_estimation/analysis_modules/estimate.py [hunk 10]
```diff
@@ -0,0 +1,407 @@
...
+        """
+        assert self._treatment_effect._patients_group in medrecord.groups, (
+            f"Patient group {self._treatment_effect._patients_group} not found in the "
+            f"data. Available groups: {medrecord.groups}"
+        )
+        assert self._treatment_effect._treatments_group in medrecord.groups, (
+            "Treatment group not found in the data. "
+            f"Available groups: {medrecord.groups}"
+        )
+        assert self._treatment_effect._outcomes_group in medrecord.groups, (
```
Comment (JabobKrauskopf) [inline medmodels/treatment_effect_estimation/analysis_modules/estimate.py:44]: these assertions should probably be ValueErrors as per the google python styleguide https://google.github.io/styleguide/pyguide.html#244-decision

Comment (JabobKrauskopf) [reply medmodels/treatment_effect_estimation/analysis_modules/estimate.py:44]: the Raises: Docstrings would need to be adjusted after changing this

File: medmodels/treatment_effect_estimation/analysis_modules/report.py [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+        methods = inspect.getmembers(
+            self._treatment_effect.estimate, predicate=inspect.ismethod
+        )
+        for name, method in methods:
+            if not name.startswith("_") | name.startswith("subject"):
+                try:
+                    results[name] = method(medrecord=medrecord)
+                except Exception as e:
+                    results[name] = f"Error: {e}"
+        return results
```
Comment (JabobKrauskopf) [inline medmodels/treatment_effect_estimation/analysis_modules/report.py:35]: this is just a fancy way of calling a few functions but with basically no readability which could break in unexpected ways without warning.

Just call the different functions and make the code readable and more importantly maintainable

File: medmodels/treatment_effect_estimation/builder.py [hunk 12]
```diff
@@ -0,0 +1,312 @@
...
+    matching_method: Optional[MatchingMethod]
+    matching_essential_covariates: Optional[MedRecordAttributeInputList]
+    matching_one_hot_covariates: Optional[MedRecordAttributeInputList]
+    matching_model: Optional[Model]
+    matching_distance_metric: Optional[Metric]
+    matching_number_of_neighbors: Optional[int]
+    matching_hyperparam: Optional[Dict[str, Any]]
+
+    def __init__(self) -> None:
+        pass
```
Comment (JabobKrauskopf) [inline medmodels/treatment_effect_estimation/builder.py:48]:
```suggestion
```

I don't think this is necessary if it does nothing

File: medmodels/treatment_effect_estimation/builder.py [hunk 13]
```diff
@@ -0,0 +1,312 @@
...
+
+    def set_treatment(self, treatment: Group) -> TreatmentEffectBuilder:
```
Comment (JabobKrauskopf) [inline medmodels/treatment_effect_estimation/builder.py:50]: these functions would more tipically be called `with_treatment`. Same goes for the other `set_` functions

File: medmodels/treatment_effect_estimation/builder.py [hunk 14]
```diff
@@ -0,0 +1,312 @@
...
+
+        Returns:
+            TreatmentEffectBuilder: The current instance of the TreatmentEffectBuilder
+                with updated time attribute.
+        """
+        self.filter_controls_operation = operation
+
+        return self
+
+    def adjust_with_propensity_matching(
```
Comment (JabobKrauskopf) [inline medmodels/treatment_effect_estimation/builder.py:225]: would probably rename this to `with_propensity_adjustment` but have no hard feelings about it. If you feel like this is more concise then thats fine by me

File: medmodels/treatment_effect_estimation/analysis_modules/matching/matching.py [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+
+MatchingMethod: TypeAlias = Literal["propensity", "nearest_neighbors"]
+
+
+class Matching(metaclass=ABCMeta):
+    """
+    The Base Class for matching.
+    """
+
+    def _preprocess_data(
```
Comment (JabobKrauskopf) [inline medmodels/treatment_effect_estimation/analysis_modules/matching/matching.py:27]: this whole function should not just take all of the data out of the medrecord into a polars DataFrame to do some calculations.

Please create a issue describing the functionality needed for the MedRecord to support @MarIniOnz @LauraBoenchenLB 

For now it is fine though

File: medmodels/treatment_effect_estimation/treatment_effect.py [hunk 16]
```diff
@@ -0,0 +1,607 @@
...
+    _matching_number_of_neighbors: int
+    _matching_hyperparam: Optional[Dict[str, Any]]
+
+    def __init__(
+        self,
+        treatment: Group,
+        outcome: Group,
+    ) -> None:
+        """
+        Initializes a Treatment Effect analysis setup with specified treatments and
```
Comment (JabobKrauskopf) [inline medmodels/treatment_effect_estimation/treatment_effect.py:77]: This docstring is out of date

File: medmodels/treatment_effect_estimation/treatment_effect.py [hunk 17]
```diff
@@ -0,0 +1,607 @@
...
+        valid and available for analysis.
+
+        Args:
+            treatment (Group): The group of treatments to analyze.
+            outcome (Group): The group of outcomes to analyze.
+        """
+        TreatmentEffect._set_configuration(self, treatment=treatment, outcome=outcome)
+
+    @classmethod
+    def builder(cls) -> TreatmentEffectBuilder:
```
Comment (JabobKrauskopf) [inline medmodels/treatment_effect_estimation/treatment_effect.py:90]: needs a docstring

File: medmodels/treatment_effect_estimation/treatment_effect.py [hunk 18]
```diff
@@ -0,0 +1,607 @@
+"""This module provides a class for analyzing treatment effects in medical records.
```
Comment (JabobKrauskopf) [inline medmodels/treatment_effect_estimation/treatment_effect.py:1]: The spacing between code blocks in this file is all over the place and not coherent.
Please re-adjust the spacing to make it uniform and combine lines into coherent code blocks

File: medmodels/treatment_effect_estimation/treatment_effect.py [hunk 19]
```diff
@@ -0,0 +1,607 @@
...
+        treatments = medrecord.group(self._treatments_group)
+
+        treatments = medrecord.group(self._treatments_group)
+
+        # Create the group with all the patients that underwent the treatment
+        for treatment in treatments:
+            treated_group.update(
+                set(
+                    medrecord.select_nodes(
+                        node().index().is_in(medrecord.group(self._patients_group))
```
Comment (JabobKrauskopf) [inline medmodels/treatment_effect_estimation/treatment_effect.py:260]:
```suggestion
                        node().in_group(self._patients_group)
```

**Review by JabobKrauskopf — CHANGES_REQUESTED** [review]
Some change requests.

Mostly good job with the test coverage
image
But it seems like there are tests missing for the propensity matching

Also: The `Matching` Abstract class, `NeighborsMatching` and `PropensityMatching` should probably be moved to the `matching` module. Is there any update on this?

File: medmodels/treatment_effect_estimation/analysis_modules/estimate.py (hunk 10 above)
Comment (JabobKrauskopf) [reply medmodels/treatment_effect_estimation/analysis_modules/estimate.py:44]: there are also some `assert ...` in the `builder.py`

File: medmodels/treatment_effect_estimation/analysis_modules/estimate.py (hunk 1 above)
Comment (JabobKrauskopf) [reply medmodels/treatment_effect_estimation/analysis_modules/estimate.py:61]: there are also some other lines which are too long (mostly by one character). In this file as well as in other files. All of these lines should be a maximum of 88 characters long

File: medmodels/treatment_effect_estimation/treatment_effect.py [hunk 20]
```diff
@@ -0,0 +1,607 @@
...
+            follow_up_period_reference (Literal["first", "last"], optional): The
+                reference point for the follow-up period. Defaults to "last".
+            outcome_before_treatment_days (Optional[int], optional): The number of days
+                before the treatment to consider for outcomes. Defaults to None.
+            filter_controls_operation (Optional[NodeOperation], optional): An optional
+                operation to filter the control group based on specified criteria.
+                Defaults to None.
+            matching_method (Optional[MatchingMethod]): The method to match treatment
+                and control groups. Defaults to None.
+            matching_essential_covariates (Optional[List[MedRecordAttribute]], optional):
```
Comment (MarIniOnz) [inline medmodels/treatment_effect_estimation/treatment_effect.py:150]: Not Optional[List[MedRecordAttribute]] but: MedRecordAttributeInputList, optional. Also the description fails to indicate the default value.

File: medmodels/treatment_effect_estimation/treatment_effect.py [hunk 21]
```diff
@@ -0,0 +1,607 @@
...
+                The essential covariates to use for matching
+            matching_one_hot_covariates (Optional[List[MedRecordAttribute]], optional):
```
Comment (MarIniOnz) [inline medmodels/treatment_effect_estimation/treatment_effect.py:152]: Same as comment above

File: medmodels/treatment_effect_estimation/treatment_effect.py [hunk 22]
```diff
@@ -0,0 +1,607 @@
...
+                The one-hot covariates to use for matching
+            matching_model (Optional[Model], optional): The model to use for matching.
```
Comment (MarIniOnz) [inline medmodels/treatment_effect_estimation/treatment_effect.py:154]: Not Optional[model] but: Model, optional (there is no way this is None)

File: medmodels/treatment_effect_estimation/treatment_effect.py [hunk 23]
```diff
@@ -0,0 +1,607 @@
...
+                Defaults to "logit".
+            matching_distance_metric (Optional[Metric], optional): The distance metric
```
Comment (MarIniOnz) [inline medmodels/treatment_effect_estimation/treatment_effect.py:156]: Same as comment above

File: medmodels/treatment_effect_estimation/treatment_effect.py [hunk 24]
```diff
@@ -0,0 +1,607 @@
...
+                to use for matching. Defaults to "mahalanobis".
+            matching_number_of_neighbors (Optional[int], optional): The number of
```
Comment (MarIniOnz) [inline medmodels/treatment_effect_estimation/treatment_effect.py:158]: Same as comment above

File: medmodels/treatment_effect_estimation/treatment_effect.py [hunk 25]
```diff
@@ -0,0 +1,607 @@
...
+        assert (
+            len(treated_group) > 0
+        ), "No patients found for the treatment groups in this MedRecord"
+
+        return treated_group
+
+    def _find_outcomes(
+        self, medrecord: MedRecord, treated_group: Set[NodeIndex]
+    ) -> Tuple[Set[NodeIndex], Set[NodeIndex], Set[NodeIndex]]:
+        """Find the patients that had the outcome after the treatment and, if specified,
```
Comment (MarIniOnz) [inline medmodels/treatment_effect_estimation/treatment_effect.py:276]: The description (I know I wrote it) is a bit difficult to understand. Maybe indicate how this "if specified" works.

File: medmodels/treatment_effect_estimation/treatment_effect.py [hunk 26]
```diff
@@ -0,0 +1,607 @@
...
+        self, medrecord: MedRecord
+    ) -> Tuple[Set[NodeIndex], Set[NodeIndex], Set[NodeIndex], Set[NodeIndex]]:
+        """
+        Identifies patients who underwent treatment and experienced outcomes, and finds
+        a control group with similar criteria but without undergoing the treatment. This
+        method supports customizable criteria filtering, time constraints between
+        treatment and outcome, and optional matching of control groups to treatment
+        groups using a specified matching class.
+
+        Initializes and sorts patient groups based on treatment and outcome, applying
```
Comment (MarIniOnz) [inline medmodels/treatment_effect_estimation/treatment_effect.py:204]: Update docstrings if we are are not using Adjust class

File: medmodels/treatment_effect_estimation/treatment_effect.py [hunk 27]
```diff
@@ -0,0 +1,607 @@
...
+        did not have the outcome before the treatment.
+
+        Args:
+            medrecord (MedRecord): An instance of the MedRecord class containing patient
+                medical data.
+            treated_group (Set[NodeIndex]): A set of patient nodes that underwent the
+                treatment.
+
+        Returns:
+            Tuple[Set[NodeIndex], Set[NodeIndex], Set[NodeIndex]]: A tuple containing
```
Comment (MarIniOnz) [inline medmodels/treatment_effect_estimation/treatment_effect.py:286]: Update the description.

File: medmodels/treatment_effect_estimation/treatment_effect.py [hunk 28]
```diff
@@ -0,0 +1,607 @@
...
+        return treated_group, washout_nodes
+
+    def _find_node_in_time_window(
+        self,
+        medrecord: MedRecord,
+        node_index: NodeIndex,
+        event_node: NodeIndex,
+        start_days: int,
+        end_days: int,
+        reference: str,
```
Comment (MarIniOnz) [inline medmodels/treatment_effect_estimation/treatment_effect.py:407]: Change to Literal["first", "last], also in docstrings

File: medmodels/treatment_effect_estimation/treatment_effect.py [hunk 29]
```diff
@@ -0,0 +1,607 @@
...
+                return True
+
+        # Return False if no event happened within the time window
+        return False
+
+    def _find_reference_time(
+        self, medrecord: MedRecord, node_index: NodeIndex, reference: str
+    ) -> pd.Timestamp:
+        """
+        Determines the timestamp of the first exposure to any treatment in the
```
Comment (MarIniOnz) [inline medmodels/treatment_effect_estimation/treatment_effect.py:461]: Update dosctrings, it is not first exposure. Can be also last exposure.

File: medmodels/treatment_effect_estimation/treatment_effect.py [hunk 30]
```diff
@@ -0,0 +1,607 @@
...
+        predefined treatment list for a specified patient node. This method is crucial
+        for analyzing the temporal sequence of treatments and outcomes.
+
+        Args:
+            medrecord (MedRecord): An instance of the MedRecord class containing patient
+                medical data.
+            node_index (NodeIndex): The patient node for which to determine the first
+                treatment exposure time.
+
```
Comment (MarIniOnz) [inline medmodels/treatment_effect_estimation/treatment_effect.py:470]: Update docstrings, one variable (reference) missing. Change the typing to Literal[] also

File: medmodels/treatment_effect_estimation/treatment_effect.py [hunk 31]
```diff
@@ -0,0 +1,607 @@
...
+        Returns:
+            pd.Timestamp: The timestamp of the first treatment exposure.
+
+        Raises:
+            AssertionError: If no treatment edge with a time attribute is found for the
+                node, indicating an issue with the data or the specified treatments.
+
+        This function iterates over all treatments and finds the earliest timestamp
```
Comment (MarIniOnz) [inline medmodels/treatment_effect_estimation/treatment_effect.py:478]: Move this text to the top of the function, not after Raises

File: medmodels/treatment_effect_estimation/treatment_effect.py [hunk 32]
```diff
@@ -0,0 +1,607 @@
...
+        """
+        Identifies control groups among patients who did not undergo the specified
+        treatments. Control groups are divided into those who had the outcome
+        (control_true) and those who did not (control_false), based on the presence of
+        specified outcome codes.
+
+        Args:
+            medrecord (MedRecord): An instance of the MedRecord class containing patient
+                medical data.
+            treated_group (Set[NodeIndex]): A set of patient nodes that underwent the
```
Comment (MarIniOnz) [inline medmodels/treatment_effect_estimation/treatment_effect.py:542]: Missing control_group in docstrings

File: medmodels/treatment_effect_estimation/treatment_effect.py [hunk 33]
```diff
@@ -0,0 +1,607 @@
...
+    def _find_controls(
+        self,
+        medrecord: MedRecord,
+        control_group: Set[NodeIndex],
+        treated_group: Set[NodeIndex],
+        rejected_nodes: Set[NodeIndex] = set(),
+        filter_controls_operation: Optional[NodeOperation] = None,
+    ) -> Tuple[Set[NodeIndex], Set[NodeIndex]]:
+        """
+        Identifies control groups among patients who did not undergo the specified
```
Comment (MarIniOnz) [inline medmodels/treatment_effect_estimation/treatment_effect.py:534]: Update docstrings to indicate how control_group, reject_nodes, and filter_controls_operation applies.

**Review by MarIniOnz — COMMENTED** [review]
Very nice PR! Excited for this work.
Commented on minor docstrings and explanations within the code

File: medmodels/treatment_effect_estimation/analysis_modules/matching/propensity.py [hunk 34]
```diff
@@ -0,0 +1,121 @@
...
+from medmodels.matching.algorithms.classic_distance_models import nearest_neighbor
+from medmodels.matching.algorithms.propensity_score import Model, calculate_propensity
+from medmodels.matching.metrics import Metric
+from medmodels.medrecord.types import MedRecordAttributeInputList, NodeIndex
+from medmodels.treatment_effect_estimation.analysis_modules.matching.matching import (
+    Matching,
+)
+
+
+class PropensityMatching(Matching):
```
Comment (MarIniOnz) [inline medmodels/treatment_effect_estimation/analysis_modules/matching/propensity.py:18]: Missin propensity matching tests

File: medmodels/treatment_effect_estimation/tests/test_treatment_effect.py [hunk 35]
```diff
@@ -0,0 +1,644 @@
...
+
+        tee = (
+            TreatmentEffect.builder()
+            .set_treatment("Rivaroxaban")
+            .set_outcome("Stroke")
+            .finish()
+        )
+
+        # Calculate metrics
+        self.assertAlmostEqual(tee.estimate.absolute_risk(self.medrecord), 1 / 6)
```
Comment (MarIniOnz) [inline medmodels/treatment_effect_estimation/tests/test_treatment_effect.py:609]: Missing the cases in which there are errors raised for these metrics. They are very difficult to replicate one by one so maybe that can be skipped? @JabobKrauskopf

**Review by MarIniOnz — COMMENTED** [review]
After a closer look, there are some missing tests, check all the raises in the treatment effect and in case some of them can be simulated, please do!
- _find_treated_patients: What if no patients are in the treatment groups?
- Same for _find_outcomes.
- _find_in_node_in_time_window: Time attribute not found in the edes attributes.
- _find_reference_time: Time attribute are not found in the edges attributes, No treatments found in the group (maybe a bit excessive checking for these). To check: when we do not find any reference time.
- find_controls: check valueErrors.
- 
-

File: medmodels/treatment_effect_estimation/tests/test_treatment_effect.py (hunk 35 above)
Comment (JabobKrauskopf) [reply medmodels/treatment_effect_estimation/tests/test_treatment_effect.py:609]: fine by me :)

File: medmodels/matching/matching_classes/__init__.py
Comment (JabobKrauskopf) [inline medmodels/matching/matching_classes/__init__.py:1]: `matching_classes` is not a good module name. I'd probably just delete the module and move all files directly into the `matching` folder

File: medmodels/treatment_effect_estimation/analysis_modules/__init__.py
Comment (JabobKrauskopf) [inline medmodels/treatment_effect_estimation/analysis_modules/__init__.py:1]: `analysis_modules` is also not a good module name. I'd just delete the module and move all files directly into the `treatment_effect_estimation` folder

File: medmodels/treatment_effect_estimation/treatment_effect.py [hunk 36]
```diff
@@ -0,0 +1,630 @@
...
+                to use for matching. Defaults to "mahalanobis".
+            matching_number_of_neighbors (int, optional): The number of
+                neighbors to match for each treated subject. Defaults to 1.
+            matching_hyperparam (Optional[Dict[str, Any]], optional): The
+                hyperparameters for the matching model. Defaults to None.
+        """
+        treatment_effect._patients_group = patients_group
+        treatment_effect._time_attribute = time_attribute
+
+        # Add the treatments and outcomes to the class
```
Comment (JabobKrauskopf) [inline medmodels/treatment_effect_estimation/treatment_effect.py:166]: unnecessary comment

File: medmodels/treatment_effect_estimation/treatment_effect.py [hunk 37]
```diff
@@ -0,0 +1,630 @@
...
+        treatment_effect._treatments_group = treatment
+        treatment_effect._outcomes_group = outcome
+
+        # configure treatment effect
```
Comment (JabobKrauskopf) [inline medmodels/treatment_effect_estimation/treatment_effect.py:170]: unnecessary comment

File: medmodels/treatment_effect_estimation/treatment_effect.py [hunk 38]
```diff
@@ -0,0 +1,630 @@
...
+        treatment_effect._washout_period_days = washout_period_days
+        treatment_effect._washout_period_reference = washout_period_reference
+        treatment_effect._grace_period_days = grace_period_days
+        treatment_effect._grace_period_reference = grace_period_reference
+        treatment_effect._follow_up_period_days = follow_up_period_days
+        treatment_effect._follow_up_period_reference = follow_up_period_reference
+        treatment_effect._outcome_before_treatment_days = outcome_before_treatment_days
+        treatment_effect._filter_controls_operation = filter_controls_operation
+
+        # configure matching
```
Comment (JabobKrauskopf) [inline medmodels/treatment_effect_estimation/treatment_effect.py:180]: unnecessary comment

File: medmodels/treatment_effect_estimation/treatment_effect.py [hunk 39]
```diff
@@ -0,0 +1,630 @@
...
+
+        for outcome in outcomes:
+            nodes_to_check = set(
+                medrecord.select_nodes(
+                    node().index().is_in(medrecord.neighbors(outcome, directed=False))
+                    & node().index().is_in(list(treated_group))
+                )
+            )
+
+            # Finding the patients that had the outcome before the treatment
```
Comment (JabobKrauskopf) [inline medmodels/treatment_effect_estimation/treatment_effect.py:309]:
```suggestion
            # Find patients that had the outcome before the treatment
```

File: medmodels/treatment_effect_estimation/treatment_effect.py [hunk 40]
```diff
@@ -0,0 +1,630 @@
...
+                            outcome,
+                            start_days=-self._outcome_before_treatment_days,
+                            end_days=0,
+                            reference="first",
+                        )
+                    }
+                )
+                nodes_to_check -= outcome_before_treatment_nodes
+
+            # Finding the patients that had the outcome after the treatment
```
Comment (JabobKrauskopf) [inline medmodels/treatment_effect_estimation/treatment_effect.py:327]:
```suggestion
            # Find patients that had the outcome after the treatment
```

File: medmodels/treatment_effect_estimation/treatment_effect.py [hunk 41]
```diff
@@ -0,0 +1,630 @@
...
+        if filter_controls_operation:
+            control_group = (
+                set(medrecord.select_nodes(filter_controls_operation)) & control_group
+            )
+
+        control_group = control_group - treated_group - rejected_nodes
+        if len(control_group) == 0:
+            raise ValueError("No patients found for control groups in this MedRecord.")
+
+        # Set the control groups and the outcomes
```
Comment (JabobKrauskopf) [inline medmodels/treatment_effect_estimation/treatment_effect.py:591]: unnecessary comment

**Review by JabobKrauskopf — CHANGES_REQUESTED** [review]
Only some minor changes now :)

I also have some bigger change requests regarding tests but these will have to wait because we need to merge this PR today.

**Review by JabobKrauskopf — APPROVED** [review]
Amazeballs. LGTM

As per the previous comment:
> I also have some bigger change requests regarding tests but these will have to wait because we need to merge this PR today.

**Review by MarIniOnz — APPROVED** [review]

**Pull request merged**